jujugui/static/gui/src/app/init.js

```diff
@@ -0,0 +1,1590 @@
+      baseURL = `${window.location.origin}${baseURL}`;
+    }
+    /**
+      Application instance of State.
+      @type {Object}
+    */
+    this.state = this._setupState(baseURL);
+
+    this._setupControllerAPI();
+
+    this._setupRomulusServices(config, window.jujulib);
+
+    /**
+      Application instance of the bundle importer.
+      @type {Object}
+    */
+    this.bundleImporter = new yui.juju.BundleImporter({
+      db: this.db,
+      modelAPI: this.modelAPI,
+      getBundleChanges: this.controllerAPI.getBundleChanges.bind(
+        this.controllerAPI),
+      makeEntityModel: yui.juju.makeEntityModel,
+      charmstore: this.charmstore,
+      hideDragOverNotification: this._hideDragOverNotification.bind(this)
+    });
+
+    if (config.gisf) {
+      document.body.classList.add('u-is-beta');
+    }
+
+    /**
+      Application instance of the ACL.
+      @type {Object}
+    */
+    this.acl = new yui.juju.generateAcl(this.controllerAPI, this.modelAPI);
+    // Listen for window unloads and trigger the unloadWindow function.
+    window.onbeforeunload = yui.juju.views.utils.unloadWindow.bind(this);
+
+    this._handleMaasServer();
+    // Feed environment changes directly into the database.
+    this._domEventHandlers['onDeltaBound'] = this.db.onDelta.bind(this.db);
+    document.addEventListener('delta', this._domEventHandlers['onDeltaBound']);
+
+    this.db.machines.after(
+        ['add', 'remove', '*:change'],
+        this.onDatabaseChanged, this);
+    this.db.services.after(
+        ['add', 'remove', '*:change'],
+        this.onDatabaseChanged, this);
+    this.db.relations.after(
+        ['add', 'remove', '*:change'],
+        this.onDatabaseChanged, this);
+    this.db.environment.after(
+        ['add', 'remove', '*:change'],
+        this.onDatabaseChanged, this);
+    this.db.units.after(
+        ['add', 'remove', '*:change'],
+        this.onDatabaseChanged, this);
+    this.db.notifications.after('add', this._renderNotifications, this);
+
+    // When someone wants a charm to be deployed they fire an event and we
+    // show the charm panel to configure/deploy the service.
+    this._domEventHandlers['onInitiateDeploy'] = evt => {
+      this.deployService(evt.detail.charm, evt.detail.ghostAttributes);
+    };
+    document.addEventListener(
+      'initiateDeploy', this._domEventHandlers['onInitiateDeploy']);
+
+    this._domEventHandlers['boundAppDragOverHandler'] =
+      this._appDragOverHandler.bind(this);
+    // These are manually detached in the destructor.
+    ['dragenter', 'dragover', 'dragleave'].forEach(eventName => {
+      document.addEventListener(
+        eventName, this._domEventHandlers['boundAppDragOverHandler']);
+    });
+
+    /**
+      Reference to the rendered topology.
+      @type {Object}
+    */
+    this.topology = this._renderTopology();
+
+    this.state.bootstrap();
+  }
+  /**
+    Creates a new instance of the charm store API. This method is idempotent.
```

Comment:
I wanted to keep as much logic out of the constructor and as much in methods as possible to make it easier to stub in tests if necessary. But yes I can make those changes to it.